Populate all affected records optionaly

turboMaCk opened the pull request:
default is false
closes #9

turboMaCk committed:
Populate all affected records optionaly

default is false
closes #9

turboMaCk committed:
update readme

turboMaCk committed:
prepare new release

turboMaCk commented:
Ready! @zdenal 

@renra Do you remember how we talk about possibility of returning all affected records as collection to prevent inconsistency of data on frontend? This is exactly it. With `populate: true` `move` action returns collection instead of one instance. Then `render json: Model.move(pramas[:position])` returns everything that changes. By default I want to keep it to `false` since it can affect performance a little bit.

zdenal commented on lib/active_record_florder.rb:
```diff
@@ -55,6 +55,14 @@ def step(value)
+      @populate = value
+    end
+
+    def get_populate
```
Why not in ruby style like getter/setter (eg. attr_accessible does - I know you have there default) do setter: `def populate=(value)` and getter: `def populate`.

turboMaCk replied:
`attr_accessible` is Rails method so it's not available in this class at all. I'd like to use `attr_accessor` but I can't figure out how to make this work.

zdenal replied:
I wrote `attr_accessor` as example how it is defining getter and setters. But I see you are using this style in rest of code also. I just wanted point that you probably mix 2 conventions one is eg. `set_populate(value)`, `get_populate()` and the other one is `populate=(value)`, `populate()` and you are using one from both. But lets skip this.

turboMaCk replied:
Well in fact this thing piss me off and I want refactor it. It's kind of ugly. I'm not much happy with all of that setter/getter code in here..

renra replied:
I see this code is supposed to be static so I guess that's the reason right? Normally in non-static code I would do something like

``` ruby
class ActiveRecordFlorder
  attr_accessor :populate, :step, :min_delta

  def initialize  # set the default here
    @populate = false
    @min_delta = 0.0005
    @step = 2**16
  end
end
```

turboMaCk replied:
:+1: this is next thing I want to try. And next thing I'm not much happy about is: https://github.com/turboMaCk/active-record-florder/blob/master/lib/active_record_florder/configurable.rb

zdenal commented on lib/active_record_florder/configurable.rb:
```diff
@@ -48,5 +52,9 @@ def min_position_delta
       self.class.next_position_step
     end
+
+    def move_populate_all
```
At all I don't understand `populate` and `move_populate_all` it is not really intuitive. Without README I would guess it populate order of items with some values or something else related with populating.

turboMaCk replied:
Well I'm not 100% sure that this naming is clean. If you have any suggestion please send it.

I use `populate` in public api because it keep it simple to use this. On another hand internally `move_populate_all` is used in instances to prevent conflicts with other possible values (remember that this is all "meta coding" which extends Models instances itself and we have not clue which attributes and methods they can have).

zdenal commented:
I would keep it also as `false` as FE just leave item where you drop it and we are not going to move items around if there is no issue. As we are not going to use it and it is not help us clean code or make our developing quicker in future - I have to ask - do we need this? I don't want get you angry and really appreciate your excellent work.

turboMaCk commented:
In fact implementation in our PRO prolatform works like:
- it calculate position and set it to model => model should stay where it was droped
- request to backed is send
- after there is response from server model's position is updated

Because of this in some edge cases position of item in list should change after response from server comes back (Trello has same issue). When we will use this option to true, than after request all affected records were updated so all positions value should be consistent between FE and BE. This is goal of having this option.

On the other hand this really should happen in some edge cases. This is why by default it's off if we (or everyone else) run into this iessues and will need to really fix them, than this option is available and can be just simply turned on for all/some models.

zdenal commented:
Yes I am aware about this edge case. Do we have some more info about this edge cases? How often it's happen?

turboMaCk commented:
I believe the chance is pretty low (at least on default settings). But it also depend on case. I do not see any issue with having this option since this is isolated gem which own "blackbox magic". Also this should not affect performance when it's off (only one if statement that is skipped is added).

zdenal commented on lib/active_record_florder/base.rb:
```diff
@@ -58,6 +58,12 @@ def move(position)
       end
 
+      # populating all affected records
+      if move_populate_all
```
I would suggest `return_all_affected` as configurable property - it says more what we are config by this property.

turboMaCk replied:
Agreeee! THX

zdenal commented:
Ok I agree have this as insurance against less possible future issues as quick switch to solving it. By me if configurable property will be more intuitive in name and not confusing I am :+1: LGTM

turboMaCk committed:
change naming in private api as @zdenal suggest

turboMaCk committed:
Improve getter and setter for configuration!!

Public api for `configure` method changed
using attr_writer(s) so `=` is now required

turboMaCk commented:
configuration refactor: https://github.com/turboMaCk/active-record-florder/pull/12

renra commented on lib/active_record_florder/base.rb:
```diff
@@ -58,6 +58,13 @@ def move(position)
+      # populating all affected records
+      # only if return_all_affected_by_move option is true
+      if return_all_affected_by_move
+        affected = [] unless affected
```
Sugar: `affected ||= []`

turboMaCk replied:
:+1: thx!

turboMaCk replied:
done here: https://github.com/turboMaCk/active-record-florder/pull/12

turboMaCk committed:
syntax suggar thx @renra

renra commented:
I think it looks great!
